### PR TITLE
Auto-compaction with segment granularity retrieve incomplete segments from timeline when interval overlap

### PR DESCRIPTION
**Pull request by maytasm**
Auto-compaction with segment granularity retrieve incomplete segments from timeline when interval overlap

### Description

Auto-compaction with segment granularity retrieve incomplete segments from timeline when interval overlap.
For example, if we have segments of YEAR granularity from 2000-01-01 to 2001-01-3001 and a segment of DAY granularity from 2000-01-01 to 2000-01-02. Let's then say that we set auto compaction with a segmentGranularity of MONTH. Auto-compaction will bucket all the existing segments into the new bucket using the configured segmentGranularity (MONTH in this example). When we iterate for segments to compact we will start at the latest and in this case would be the December (2000-12-01 to 2000-12-31) bucket. We would get only the segment with YEAR granularity in this bucket. We would then create a compaction task with the interval 2000-01-01 to 2001-12-31 from the segment we just retrieved. We would also include a sha256OfSortedSegmentIds which would only be the YEAR granularity segment (2000-01-01 to 2001-12-31). However, when the compaction task runs, it would get all segment from 2000-01-01 to 2001-01-01, which would include both segments (a segment of YEAR granularity from 2000-01-01 to 2001-12-31 and a segment of DAY granularity from 2000-01-01 to 2000-01-02)

This PR fixes this problem but keeping the original timeline in the CompactibleTimelineObjectHolderCursor that can then we use to fetch the original segments back using the umbrella interval of the transformed segments.

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 1]
```diff
@@ -249,14 +246,13 @@ private void updateQueue(String dataSourceName, DataSourceCompactionConfig confi
   private static class CompactibleTimelineObjectHolderCursor implements Iterator<List<DataSegment>>
   {
     private final List<TimelineObjectHolder<String, DataSegment>> holders;
-    private final Map<Pair<Interval, SegmentId>, ShardSpec> originalShardSpecs;
-    private final Map<Pair<Interval, SegmentId>, String> originalVersion;
+    private final VersionedIntervalTimeline<String, DataSegment> originalTimeline;
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:249]: Please add `@Nullable` here too.

**Review by jihoonson — COMMENTED** [review]
+1 after CI

File: integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java [hunk 2]
```diff
@@ -438,6 +438,67 @@ public void testAutoCompactionDutyWithSegmentGranularityAndExistingCompactedSegm
...
+      verifyQuery(INDEX_QUERIES_RESOURCE);
+      verifySegmentsCompacted(1, MAX_ROWS_PER_SEGMENT_COMPACTED);
+      checkCompactionIntervals(expectedIntervalAfterCompaction);
+
+      loadData(INDEX_TASK);
+      verifySegmentsCount(5);
+      verifyQuery(INDEX_QUERIES_RESOURCE);
+      // 5 segments. 1 compacted YEAR segment and 4 newly ingested DAY segments across 2 days
+      // We wil have one segment with interval of 2013-01-01/2014-01-01 (compacted with YEAR) from the compaction
+      // two segments with interval of 2013-09-01/2013-10-01 (newly ingested with DAY)
```
Comment (jon-wei) [inline integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:472]: Is the comment here correct? The comment has monthly intervals but the ingestion spec uses day granularity

Comment (maytasm) [reply integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:472]: The interval in the comment is incorrect. It should be one day interval

Comment (maytasm) [reply integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:472]: Done

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java (hunk 1 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:249]: Done

**Review by jon-wei — APPROVED** [review]

**Pull request merged**